### PR TITLE
Auto-detect CloudLocation for Storage

### PR DESCRIPTION
**Pull request by juliusl**
Addresses #1156 and #1127

### CODE REVIEW

File: sdk/storage/src/cloud_location.rs [hunk 1]
```diff
@@ -44,20 +92,115 @@ impl CloudLocation {
...
+                            account: account.clone(),
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        _ => {
+                            todo!()
+                        }
+                    };
+                } else {
+                    // Default to PROD
```
Comment (juliusl) [inline sdk/storage/src/cloud_location.rs:147]: Should we panic here instead to fail fast?

File: sdk/storage/src/cloud_location.rs [hunk 2]
```diff
@@ -27,6 +55,26 @@ pub enum CloudLocation {
...
+    /// If either of these methods fail, this will default to public azure.
+    ///
+    /// Cloud names can be listed with, `az cloud list --output table`. Current public values are:
+    /// - AzureCloud - Public azure, in some cases may be AzurePublicCloud in env var form
+    /// - AzureChinaCloud
+    /// - AzureUSGovernment
+    /// - AzureGermanCloud - Might be deprecating based on public documentation but still shows up in `az cloud list`
+    ///
+    pub fn auto_detect(account: impl AsRef<str>, credentials: StorageCredentials) -> CloudLocation {
+        CloudLocation::AutoDetect {
```
Comment (juliusl) [inline sdk/storage/src/cloud_location.rs:72]: Alternatively, could add a fallback field, to handle test/custom clouds?

File: sdk/storage/src/cloud_location.rs [hunk 3]
```diff
@@ -107,11 +270,34 @@ impl TryFrom<&Url> for CloudLocation {
...
+                && url.has_host()
+                && url.port().is_some() =>
+            {
+                if let Some(host) = url.host() {
+                    match host {
+                        url::Host::Ipv4(ip) => Ok(CloudLocation::Emulator {
+                            address: format!("{ip}"),
+                            port: url.port().expect("should have a port"),
+                        }),
+                        _ => Err(azure_core::Error::with_message(
```
Comment (juliusl) [inline sdk/storage/src/cloud_location.rs:287]: This is an assumption, but could as well just pass the other cases as well

File: sdk/storage/src/cloud_location.rs [hunk 4]
```diff
@@ -44,28 +108,127 @@ impl CloudLocation {
...
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        AZURE_GERMAN_CLOUD => CloudLocation::GermanCloud {
+                            account: account.clone(),
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        _ => {
+                            todo!()
```
Comment (rylev) [inline sdk/storage/src/cloud_location.rs:160]: Return `Err`?

File: sdk/storage/src/cloud_location.rs [hunk 5]
```diff
@@ -44,28 +108,127 @@ impl CloudLocation {
...
+                .join(".azure/config")
+                .canonicalize()
+                .ok()
+                .and_then(|config| File::open(config).ok())
+            {
+                let mut lines = BufReader::new(config).lines();
+
+                while let Some(Ok(line)) = lines.next() {
+                    // Alternatively, import serde_toml for parsing this file, but probably better off doing
+                    // that by creating a struct dedicated to managing the config file
```
Comment (rylev) [inline sdk/storage/src/cloud_location.rs:208]: we should either do this or remove this comment. I'm unsure if the dependency is worth it if we can keep things this short.

File: sdk/storage/src/cloud_location.rs [hunk 6]
```diff
@@ -44,28 +108,127 @@ impl CloudLocation {
...
+                    if line.trim() == "[cloud]" {
+                        if let Some(Ok(name)) = lines.next() {
```
Comment (rylev) [inline sdk/storage/src/cloud_location.rs:210]: Is the next line allowed to be empty?

File: sdk/storage/src/cloud_location.rs [hunk 7]
```diff
@@ -169,6 +358,184 @@ mod tests {
...
+                .url(ServiceType::Blob)
+                .expect("should return a url")
+                .as_str(),
+            "https://test_account.blob.core.cloudapi.de/"
+        );
+
+        std::env::remove_var("AZURE_CLOUD_NAME");
+
+        std::env::set_var("HOME", ".test");
+        std::fs::create_dir_all(".test/.azure").expect("should be able to make test dir");
```
Comment (rylev) [inline sdk/storage/src/cloud_location.rs:450]: Should we be writing these into a temp folder?

**Review by rylev — CHANGES_REQUESTED** [review]

File: sdk/storage/src/cloud_location.rs (hunk 6 above)
Comment (juliusl) [reply sdk/storage/src/cloud_location.rs:181]: I'm not sure, I think this file is either set by az cli, or when Azure Compute initializes the VM

**Comment by juliusl** [comment]
@rylev Addressed all comments

File: sdk/storage/src/cloud_location.rs (hunk 6 above)
Comment (juliusl) [reply sdk/storage/src/cloud_location.rs]: Ok I think I understand what you might be concerned about, added some more resiliency.

File: sdk/storage/src/cloud_location.rs [hunk 8]
```diff
@@ -44,28 +88,123 @@ impl CloudLocation {
...
+            | CloudLocation::AutoDetect { credentials, .. } => credentials,
             CloudLocation::Emulator { .. } => &EMULATOR_CREDENTIALS,
-            CloudLocation::Custom { credentials, .. } => credentials,
+        }
+    }
+
+    /// Finds the cloud name, first by environment variable, then by parsing the current user's $HOME/.azure/config file
+    ///
+    fn find_cloud_name() -> Option<String> {
+        if let Ok(name) = std::env::var("AZURE_CLOUD_NAME") {
```
Comment (cataggar) [inline sdk/storage/src/cloud_location.rs:167]: Is `AZURE_CLOUD_NAME` used by any existing Azure SDKs? https://github.com/Azure/azure-sdk-for-net/search?q=AZURE_CLOUD_NAME comes up empty.

File: sdk/storage/src/cloud_location.rs [hunk 9]
```diff
@@ -44,28 +88,123 @@ impl CloudLocation {
...
+            Some(name)
+        } else if let Ok(home_dir) = std::env::var("HOME") {
+            if let Some(config) = PathBuf::from(home_dir)
+                .join(".azure/config")
```
Comment (cataggar) [inline sdk/storage/src/cloud_location.rs:171]: I'm not sure the storage crate should be reading `$HOME/.azure/config`. Is something like this done in the other Azure SDKs?

**Review by cataggar — CHANGES_REQUESTED** [review]

File: sdk/storage/src/cloud_location.rs (hunk 8 above)
Comment (juliusl) [reply sdk/storage/src/cloud_location.rs:167]: I couldn't find anything either, but apparently, it's used by other libraries for cloud detection. I actually think this env variable might make more sense, https://github.com/Azure/azure-sdk-for-net/blob/eca8faf3c868bbb75ed8a104661bc77f5a976c9a/sdk/identity/Azure.Identity/src/EnvironmentVariables.cs#L30

File: sdk/storage/src/cloud_location.rs (hunk 9 above)
Comment (juliusl) [reply sdk/storage/src/cloud_location.rs:171]: Can't find any other usage besides the cli repos. It appears the storage sdk for .net handles this by using an 'EndpointSuffix=' arg in the connection string. Is there a possible middle ground of moving this to a shared location or what are your thoughts?

**Comment by cataggar** [comment]
I think we should follow the lead of the other official Azure SDKs here.

**Pull request closed without merging**